[LV] Optimise users of induction variables in early exit blocks

david-arm opened the pull request:
This is the second of two PRs that attempts to improve the IR
generated in the exit blocks of vectorised loops with uncountable
early exits. It follows on from PR #128880. In this PR I am
improving the generated code for users of induction variables in
early exit blocks.

This required using a newly add VPInstruction called
FirstActiveLane, which calculates the index of the first active
predicate in the mask operand.

I have added a new function optimizeEarlyExitInductionUser that
is called from optimizeInductionExitUsers when handling users in
early exit blocks.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -882,6 +882,8 @@ class VPInstruction : public VPRecipeWithIRFlags,
     // Extracts the first active lane of a vector, where the first operand is
     // the predicate, and the second operand is the vector to extract.
     ExtractFirstActive,
+    // Calculates the first active lane index of the vector predicate operand.
+    FirstActiveLane,
```
Do we need both `ExtractFirstActive` and `FirstActiveLane` or can we use FirstActiveLane with a regular extract VPInstruction?

david-arm replied:
Yep, I can do that. Split off into https://github.com/llvm/llvm-project/pull/131118.

david-arm commented:
Gentle ping. :)

fhahn commented on llvm/test/Transforms/LoopVectorize/single_early_exit_live_outs.ll:
```diff
@@ -86,6 +85,427 @@ loop.end:
 }
 
 
+define i32 @same_exit_block_pre_inc_use1_iv64_endi32_step2() {
```
That looks fine, thanks

fhahn reviewed (approved):
LGTM, thanks

david-arm committed:
Add tests

david-arm committed:
[LV] Optimise users of induction variables in early exit blocks

This is the second of two PRs that attempts to improve the IR
generated in the exit blocks of vectorised loops with uncountable
early exits. It follows on from PR #128880. In this PR I am
improving the generated code for users of induction variables in
early exit blocks.

This requires using a newly added VPInstruction called
FirstActiveLane, which calculates the index of the first active
predicate in the mask operand.

I have added a new function optimizeEarlyExitInductionUser that
is called from optimizeInductionExitUsers when handling with
users in early exit blocks.

david-arm committed:
Rebase + fix

* Regenerated some new tests after rebase.
* Fixed an issue where we were assuming that FirstActiveLane
always returned a 64-bit value.

david-arm commented:
Rebased. Ran make check-all locally and all tests passed, except two clang tests that are unrelated to this PR.
